### PR TITLE
Do not discard byref zeros in `native int + byref` additions

### LINKED ISSUE #67600: Test failure: Assertion failed 'addr->TypeIs(TYP_BYREF) || addr->IsIntegralConst(0)'
in `runtime-coreclr libraries-jitstress` pipeline

For example,
```
D:\h\w\BC950A38\w\99FA088B\e>set COMPlus 
COMPlus_JitStress=1
COMPlus_TieredCompilation=0

D:\h\w\BC950A38\w\99FA088B\e>call RunTests.cmd --runtime-path D:\h\w\BC950A38\p 
----- start Tue 04/05/2022  2:06:59.25 ===============  To repro directly: ===================================================== 
pushd D:\h\w\BC950A38\w\99FA088B\e\
"D:\h\w\BC950A38\p\dotnet.exe" exec --runtimeconfig System.Xml.RW.CharCheckingReader.Tests.runtimeconfig.json --depsfile System.Xml.RW.CharCheckingReader.Tests.deps.json xunit.console.dll System.Xml.RW.CharCheckingReader.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================

D:\h\w\BC950A38\w\99FA088B\e>"D:\h\w\BC950A38\p\dotnet.exe" exec --runtimeconfig System.Xml.RW.CharCheckingReader.Tests.runtimeconfig.json --depsfile System.Xml.RW.CharCheckingReader.Tests.deps.json xunit.console.dll System.Xml.RW.CharCheckingReader.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Xml.RW.CharCheckingReader.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Xml.RW.CharCheckingReader.Tests (found 1 test case)
  Starting:    System.Xml.RW.CharCheckingReader.Tests (parallel test collections = on, max threads = 8)

Assert failure(PID 3644 [0x00000e3c], Thread: 2220 [0x08ac]): Assertion failed 'addr->TypeIs(TYP_BYREF) || addr->IsIntegralConst(0)' in 'System.Xml.Tests.TCErrorCondition:v1():int:this' during 'Morph - Global' (IL size 522; hash 0xa5956e47; FullOpts)

    File: D:\a\_work\1\s\src\coreclr\jit\gentree.h Line: 5791
    Image: D:\h\w\BC950A38\p\dotnet.exe
```

https://dev.azure.com/dnceng/public/_build/results?buildId=1699992&view=ms.vss-test-web.build-test-results-tab&runId=46335470&resultId=184901&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

Seen in `net7.0-windows-Release-arm64-CoreCLR_checked-jitstress1-Windows.10.Arm64v8.Open`:
- `System.Runtime.Tests Work Item`
- `System.Xml.RW.CharCheckingReader.Tests Work Item`
- more

Also, `net7.0-Linux-Release-arm-CoreCLR_checked-jitstress2`: `System.Numerics.Vectors.Tests Work Item`, `System.Runtime.Tests Work Item`

Also, `net7.0-Linux-Release-arm-CoreCLR_checked-jitstress1_tiered` and more.

@SingleAccretion PTAL
cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by SingleAccretion**
This should:

a) Ensure we're not "losing" the fact `ADD` produces a newly-born GC value (a mostly theoretical issue).
b) Reinstate the assert from #67600.

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1720601&view=ms.vss-build-web.run-extensions-tab).

Fixes #67600.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib I would like to request a run of `runtime-coreclr libraries-jitstress` for this change.

**Comment by SingleAccretion** [comment]
Libraries stress "passed", so the at least in the first approximation this fix should be sufficient.

Failures:
1) `Linux_musl arm64 Debug`: looks related to #67031.
2) [`OSX x64 checked`](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-68086-merge-feedec91800c450cac/PayloadGroup0/1/console.3c65db67.log?sv=2019-07-07&se=2022-05-05T21%3A56%3A54Z&sr=c&sp=rl&sig=SKfsH4hpmc9BqZ%2F9Ql7dNdEIZ1mQ2oqpNcC814vMTnw%3D): no existing issue, but does not look related (perhaps could be re-run).
3) Libraries stress, [`Linux arm Release`](https://helix.dot.net/api/2019-06-17/jobs/25587e7b-15a8-4375-87dc-9d5188503df6/workitems/System.Text.Encodings.Web.Tests/console): #67046.
4) Libraries stress, [`windows x86 Release`](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-68086-merge-b8193db142f9459fa5/System.Text.Json.SourceGeneration.Roslyn4.0.Tests/1/console.7534006a.log?sv=2019-07-07&se=2022-05-06T08%3A59%3A23Z&sr=c&sp=rl&sig=3mHeqtzhZRYgVaHMy05l0T%2BkBSOfF0j%2BO6gmzQsE%2F3A%3D): #67761.
5) Libraries stress, `windows arm64 Release`:
     - [VM assert](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-68086-merge-ba767eac66b545a18a/System.Collections.Tests/1/console.8af4df07.log?sv=2019-07-07&se=2022-05-06T09%3A07%3A46Z&sr=c&sp=rl&sig=4d6AF6uZG0l6n6nao%2FbBoj%2FJVAN9aXnSmzdutOWBcq4%3D): looks like #67083.
     - [UnauthorizedAccessException](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-68086-merge-f50072925edf4edf9a/System.Diagnostics.EventLog.Tests/1/console.58c362ee.log?sv=2019-07-07&se=2022-05-06T09%3A07%3A46Z&sr=c&sp=rl&sig=u3fiN0Plly4oQLpJ4savfHjGGxoVdB6hUzmk2jLoe2Q%3D): pre-existing - [earlier run](https://dev.azure.com/dnceng/public/_build/results?buildId=1721515&view=logs&j=43e320c8-9861-5cff-9f08-1de31d3a4445&t=68d78252-9086-548a-668f-04eaeddce911).
     - Timeout: pre-existing - [earlier run](https://dev.azure.com/dnceng/public/_build/results?buildId=1721515&view=results).

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -13114,9 +13114,8 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
         DEBUG_DESTROY_NODE(constTwo);
     }
 
-    // Fold (x + 0) - given it won't change the tree type to TYP_REF.
-    // TODO-Bug: this code will lose the GC-ness of a tree like "native int + byref(0)".
-    if (op2->IsIntegralConst(0) && ((add->TypeGet() == op1->TypeGet()) || !op1->TypeIs(TYP_REF)))
+    // Fold (x + 0) - given it won't change the tree type.
+    if (op2->IsIntegralConst(0) && (genActualType(add) == genActualType(op1)))
```
Comment (sandreenko) [inline src/coreclr/jit/morph.cpp:13118]: Could we have something like TYP_LONG(N) + TYP_INT(0)?

Comment (SingleAccretion) [reply src/coreclr/jit/morph.cpp:13118]: Technically that is invalid IR, practically we may see it (due to bugs / invalid IL in CLR tests / both).

I suppose this code should be correct for that case (well, not if it is `long + int => int`, but then that's actually invalid and I have never seen anything produce such forms).

Comment (TIHan) [reply src/coreclr/jit/morph.cpp:13077]: I don't think we will see such forms either, as you said, it is technically invalid.

**Review by TIHan — APPROVED** [review]
LGTM.

**Pull request merged**